### ISSUE TITLE
Dataset.iterbatches not iterating in batch_size 2

### ISSUE THREAD

**Issue by proteneer** [deepchem/deepchem#911]
So I'm running into this again but under a different circumstance. This is different from #881 

If shard_size isn't a multiple of batch_size, then iterbatches once again returns a different number of batches.

Eg.:

10 shards of 15 elements each = 150 items

batch_size = 14

expected number of batches = ceil(150/14) == 11

but instead we ~22 batches because the remainders isn't carried over.

**Comment by rbharath** [comment]
Could you make a small reproducing code segment? It's not clear to me how we get 22 batches here

**Comment by proteneer** [comment]
Add these to test_datasets.py

``` python
  def test_disk_iterate_batch_size_uneven_shards(self):

    batch_size = 5
    shard_sizes = [21, 11, 41, 21, 51]

    def shard_generator():
      for sz in shard_sizes:
        X_b = np.ones((sz, 1))
        y_b = np.ones((sz, 1))
        w_b = np.ones((sz, 1))
        ids_b = np.ones((sz, 1))
        yield X_b, y_b, w_b, ids_b

    dataset = dc.data.DiskDataset.create_dataset(
      shard_generator())

    total_size = sum(shard_sizes)

    assert dataset.X.shape[0] == total_size

    for bidx, (X_b, _, _, _) in enumerate(dataset.iterbatches(
      batch_size=batch_size,
      pad_batches=False,
      deterministic=True)):
      pass

    print(bidx, math.ceil(total_size/batch_size))

    assert bidx == math.ceil(total_size/batch_size)

  def test_disk_iterate_batch_size_even_shards(self):

    batch_size = 5
    shard_sizes = [31, 31, 31, 31, 31]

    def shard_generator():
      for sz in shard_sizes:
        X_b = np.ones((sz, 1))
        y_b = np.ones((sz, 1))
        w_b = np.ones((sz, 1))
        ids_b = np.ones((sz, 1))
        yield X_b, y_b, w_b, ids_b

    dataset = dc.data.DiskDataset.create_dataset(
      shard_generator())

    total_size = sum(shard_sizes)

    assert dataset.X.shape[0] == total_size

    for bidx, (_, _, _, _) in enumerate(dataset.iterbatches(
      batch_size=batch_size,
      pad_batches=False,
      deterministic=True)):
      pass

    print(bidx, math.ceil(total_size/batch_size))

    assert bidx == math.ceil(total_size/batch_size)

```

First assertion is 34, 31, and second assertion fails with 32, 29

**Comment by proteneer** [comment]
curious to hear @peastman thoughts on this

**Comment by proteneer** [comment]
Note that this is quite important because a lot of the multithreaded code for queueing and running sessions requires us to pre-compute the number of expected enqeueue and run ops a priori. I also view the number of shards etc. as an implementation detail, so I'd expect remainder carrying to occur automatically.

**Comment by lilleswing** [comment]
How would you propose this?  Its a function on dataset?

```python
Dataset.get_batchs(batch_size)
```

**Comment by proteneer** [comment]
we can either make a new iter_exact_batches that guarantees this or modify the current one so it has the correct behavior

**Comment by proteneer** [comment]
PS I can't see any reason why we shouldn't do the carry-over automatically, aside from requiring up to O(max(batch_size, shard_size)) additional space.

**Comment by proteneer** [comment]
Note that this would also change the behavior so setting pad_batches=True in iterbatches would only pad the *last* global batch across all shards, and not the last local batch within each shard.

**Comment by proteneer** [comment]
after talking with @lilleswing off-line we're convinced that this is the right behavior. If @rbharath and @peastman agree I will submit a PR to fix this.

**Comment by rbharath** [comment]
This seems pretty reasonable to me as well. Agree that shard sizes are an implementation detail that shouldn't be externally visible to degree possible.

**Comment by rbharath** [comment]
Closing since I think solved by #914.
